# eBay Kleinanzeigen Car Sales
### Objective - To clean and analyze used car sales data from the german equivalent of eBay motors. 

The data dictionary provided with data is as follows:

- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [48]:
import pandas as pd

## Import

data file stored on GitHub

In [49]:
autos = pd.read_csv('https://raw.githubusercontent.com/aaronsang/guided-data-science-projects/master/data/autos.csv', encoding='Latin-1')

## Data Evaluation & Cleaning

Data contains 20 columns, 50,000 rows, mostly string data with some integers. Some column names are Camel Case. Some column names are not descriptive enough.

### Cleaning Todo:

- [x] Standardize casing across column names
- [x] Rename columns whos names are not descriptive enough
- [x] Clean price column: (convert to int, remove outliers)
- [x] Clean odometer column: (convert to int, remove outliers)
- [x] Identify and remediate any columns with null values
  - [x] vehicle_type
  - [x] gearbox
  - [x] model
  - [x] fuel_type
  - [x] unrepaired_damage


In [50]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [51]:
autos.describe(include="all")

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [52]:
autos.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [53]:
# Modified column names from CamelCase to snake_case
autos.columns
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_columns

In [54]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

I renamed a few columns so that they more accurately describe their data points and to create snake-case consistency.

- yearOfRegistration to registration_year
- monthOfRegistration to registration_month
- notRepairedDamage to unrepaired_damage
- dateCreated to ad_created

In [55]:
# Coverted odometer_km column to integer and removed comma and km label
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype('int64')
# removed $ and ,; converted to int
autos['price'] = autos['price'].str.replace('$',"").str.replace(',',"").astype('int64')
# Renamed odometer column to odometer_km
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)


In [56]:
# Verification of column renaming and dtype
autos[['price','odometer_km']].dtypes

price          int64
odometer_km    int64
dtype: object

In [57]:
autos['price'].unique().shape[0]

2357

In [58]:
print(autos['price'].describe())
print()
print(autos['price'].value_counts())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64


In [59]:
# Evaluation of low prices
print(autos['price'].value_counts().sort_index(ascending=True).head())
# Evaluation of high prices
print(autos['price'].value_counts().sort_index(ascending=False).head(15))

0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64


In [60]:
low_prices = autos.loc[autos['price'] < 1].shape[0]
high_prices = autos.loc[autos['price'] > 350_000].shape[0]
print(f"Rows with prices < 1: {low_prices}")
print(f"Rows with prices > 350k: {high_prices}")
print(f"low and high prices combined: {low_prices + high_prices}")

Rows with prices < 1: 1421
Rows with prices > 350k: 14
low and high prices combined: 1435


There are 1421 rows with their prices < 1. Looking at the maximum prices, it jumps from 350k to 999990. Since there are such a small amount of rows with prices >350k, I will remove the prices under <1, and <350k.

In [61]:
autos = autos[autos['price'].between(1, 350_000)]
print(autos.shape)

(48565, 20)


The sum of low and high prices I removed matches the difference between the starting index of 50k and the amount of rows remaining.

In [62]:
autos['odometer_km'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

Reviewing the odometer values, doesn't appear to have any outliers and the data looks well rounded. Most of the vehicles have 150,000 km.

In [63]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Looks like the dates are the first 10 characters. I'll extract just the date to evaluate the distribution of dates.

In [64]:
date_crawled = autos['date_crawled'].str[:10]
date_crawled.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [65]:
ad_created = autos['ad_created'].str[:10]
ad_created.value_counts(normalize=True,dropna=False).head(20) * 100

2016-04-03    3.885514
2016-03-20    3.794914
2016-03-21    3.757850
2016-04-04    3.685782
2016-03-12    3.675486
2016-03-14    3.518995
2016-04-02    3.514877
2016-03-28    3.498404
2016-03-07    3.473695
2016-03-29    3.403686
2016-03-15    3.401627
2016-03-19    3.368681
2016-04-01    3.368681
2016-03-30    3.350149
2016-03-08    3.331617
2016-03-09    3.315145
2016-03-11    3.290435
2016-03-22    3.280140
2016-03-26    3.226604
2016-03-23    3.206013
Name: ad_created, dtype: float64

In [66]:
last_seen = autos['last_seen'].str[:10]
last_seen.value_counts(normalize=True, dropna=False).head(20) * 100

2016-04-06    22.180583
2016-04-07    13.194688
2016-04-05    12.476063
2016-03-17     2.808607
2016-04-03     2.520334
2016-04-02     2.491506
2016-03-30     2.477093
2016-04-04     2.448265
2016-03-12     2.378256
2016-03-31     2.378256
2016-04-01     2.279419
2016-03-29     2.234119
2016-03-22     2.137342
2016-03-28     2.085864
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-23     1.853186
2016-03-26     1.680222
Name: last_seen, dtype: float64

Appears most of the data was obtained throughout the month of March and is spread pretty evenly. However the first week of April seems to be the most popular viewing time.

In [67]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [68]:
(~autos['registration_year'].between(1900,2016)).sum() / len(autos) * 100

3.8793369710697

Since the the rows outside of 1900 - 2016 only makes up 3.8% of the dataset, I will remove these.

In [69]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True).head(15) * 100

2000    6.760781
2005    6.289497
1999    6.205951
2004    5.790364
2003    5.781796
2006    5.719672
2001    5.646837
2002    5.325507
1998    5.062017
2007    4.877788
2008    4.744971
2009    4.466485
1997    4.179431
2011    3.476789
2010    3.403954
Name: registration_year, dtype: float64

Most of the vehicles were registered within the past 20 years or so.

In [70]:
brands = autos['brand'].value_counts(normalize=True) * 100
print(brands)

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

Based on the results VW, BMW, and Opel since combined they make up slightly over 40% of the dataset with VW leading in sales by far.

In [77]:
top_brands = brands[brands > 2.0].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')

In [78]:
brand_mean_prices = {}
for brand in top_brands:
    brand_series = autos[autos['brand'] == brand]
    mean_price = brand_series['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813}

I chose to aggregate by the brands that consist of more than 2% of the data set. Audis seem to have the highest sell prices followed closely by mercedes then bmw. German vehicles seem to have the highest sell prices in....Germany.

Fiat, opel, and peugeot seem to be the least valuable.

Evaluating the mean mileage between these brands may yield some greater insights.

In [80]:
top_brand_mileage = {}
for brand in top_brands:
    brand_series = autos[autos['brand'] == brand]
    mean_mileage = brand_series['odometer_km'].mean()
    top_brand_mileage[brand] = mean_mileage

top_brand_mileage

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056,
 'renault': 128071.33121308497,
 'peugeot': 127153.62526920316,
 'fiat': 117121.9715956558}

In [98]:
brand_price_series = pd.Series(brand_mean_prices)
mileage_series = pd.Series(top_brand_mileage)
prices_and_mileage = pd.DataFrame(brand_price_series,columns=['mean_price'])
prices_and_mileage['mean_mileage'] = mileage_series
prices_and_mileage.sort_values('mean_price',ascending=False)

mean_price   mean_mileage
audi                 9336  129157.386785
mercedes_benz        8628  130788.363313
bmw                  8332  132572.513140
volkswagen           5402  128707.158791
ford                 3749  124266.012872
peugeot              3094  127153.625269
opel                 2975  129310.035842
fiat                 2813  117121.971596
renault              2474  128071.331213

Mileage doesn't seem to affect sell prices. The top brands actually seem to have higher miles.